File: 0200-number-of-islands/0200-number-of-islands.cpp
class Solution {
public:
    
    void solve(vector<vector<char>>&grid, int i, int j, vector<vector<bool>>& vis){
        int m = grid.size();
        int n = grid[0].size();
        if(i<0 or j<0 or i>=m or j>=n or vis[i][j]) return;
        
        if(grid[i][j]=='1' and !vis[i][j]){
            vis[i][j] = true;
            solve(grid, i+1, j, vis);
            solve(grid, i-1, j, vis);
            solve(grid, i, j+1, vis);
            solve(grid, i, j-1, vis);
            
        }
    }
    
    int numIslands(vector<vector<char>>& grid) {
        int ans = 0;
        vector<vector<bool>>vis(grid.size(),vector<bool>(grid[0].size(), false));
        for(int i = 0; i<grid.size(); i++){
            for(int j = 0; j<grid[0].size(); j++){
                if(grid[i][j] == '1' and !vis[i][j]){
                    vis[i][j]=true;
                    ans++;
                    solve(grid, i+1, j, vis);
                    solve(grid, i, j+1, vis);
                }
            }
        }
        return ans;
    }
};

File: 0200-number-of-islands/0200-number-of-islands.java
class Solution {
    
    public static void solve(char[][] grid, int i, int j, boolean [][] vis){
        int m = grid.length;
        int n = grid[0].length;
        
        if(i<0 || i>=m || j<0 || j>=n || vis[i][j]==true){
            return;
        }
        if(grid[i][j]=='1' && vis[i][j] == false){
            vis[i][j] = true;
            solve(grid, i+1, j, vis);
            solve(grid, i-1, j, vis);
            solve(grid, i, j+1, vis);
            solve(grid, i, j-1, vis);
        }
    }
    
    public int numIslands(char[][] grid) {
        int ans = 0;
        int m = grid.length;
        int n = grid[0].length;
        boolean [][] vis = new boolean[m][n];
        for(int i = 0; i<m; i++){
            for(int j = 0; j<n; j++){
                if(grid[i][j]=='1' && vis[i][j] == false){
                    ans++;
                    vis[i][j] = true;
                    solve(grid, i+1, j, vis);
                    solve(grid, i, j+1, vis);
                }
            }
        }
        return ans;
    }
}